lib/fuse_lowlevel.c

```diff
@@ -3059,22 +3041,35 @@ int fuse_session_receive_buf_int(struct fuse_session *se, struct fuse_buf *buf,
 				"fuse: failed to allocate read buffer\n");
 			return -ENOMEM;
 		}
+		buf->mem_size = se->bufsize;
 	}
 
 restart:
 	if (se->io != NULL) {
 		/* se->io->read is never NULL if se->io is not NULL as
 		specified by fuse_session_custom_io()*/
-		res = se->io->read(ch ? ch->fd : se->fd, buf->mem, se->bufsize,
+		res = se->io->read(ch ? ch->fd : se->fd, buf->mem, buf->mem_size ?: se->bufsize,
 					 se->userdata);
 	} else {
-		res = read(ch ? ch->fd : se->fd, buf->mem, se->bufsize);
+		res = read(ch ? ch->fd : se->fd, buf->mem, buf->mem_size ?: se->bufsize);
 	}
 	err = errno;
 
 	if (fuse_session_exited(se))
 		return 0;
 	if (res == -1) {
+		if (err == EINVAL && buf->mem_size && se->bufsize > buf->mem_size)  {
+		    free(buf->mem);
+		    buf->mem = malloc(se->bufsize);
```

Comment:
realloc would be preferable here I think, not because of the need to copy the buffer contents across but because on failure the old buffer is still allocated.

I know the changeable buffer size is new but is there any documentation or link what describes the kernel behaviour here?  At least a comment about why this is needed and when this code-path might be taken.

Finally, what's the value of bufsize here, hopefully not `UINT_MAX`?

Reply:
kernel side
```
static ssize_t fuse_dev_do_read(struct fuse_dev *fud, struct file *file,
				struct fuse_copy_state *cs, size_t nbytes)
{
	ssize_t err;
	struct fuse_conn *fc = fud->fc;
	struct fuse_iqueue *fiq = &fc->iq;
	struct fuse_pqueue *fpq = &fud->pq;
	struct fuse_req *req;
	struct fuse_args *args;
	unsigned reqsize;
	unsigned int hash;

	/*
	 * Require sane minimum read buffer - that has capacity for fixed part
	 * of any request header + negotiated max_write room for data.
	 *
	 * Historically libfuse reserves 4K for fixed header room, but e.g.
	 * GlusterFS reserves only 80 bytes
	 *
	 *	= `sizeof(fuse_in_header) + sizeof(fuse_write_in)`
	 *
	 * which is the absolute minimum any sane filesystem should be using
	 * for header room.
	 */
	if (nbytes < max_t(size_t, FUSE_MIN_READ_BUFFER,
			   sizeof(struct fuse_in_header) +
			   sizeof(struct fuse_write_in) +
			   fc->max_write))
		return -EINVAL;


```

Reply:
The kernel max_write value will stay constant per connection (even if the sysctl value changes). The libfuse buffer must be large enough to accommodate max_write or the read will error out with -EINVAL.

Reply:
Is there anyway to read the value of max_write from userspace or have I missed that code somewhere?

Reply:
The sysctl value is for max_pages, whihc you can read from userspace through `/proc/sys/fs/fuse/max_pages_limit`, and the kernel internally uses max_pages to determine max_write.

Reply:
> Finally, what's the value of bufsize here, hopefully not UINT_MAX?

The value of bufsize here is the max write value that the user set in their init callback through "se->conn.max_write = x"